compiler/src/iree/compiler/Dialect/VM/IR/VMOpFolders.cpp

```diff
@@ -579,41 +579,48 @@ static Attribute constFoldFloatUnaryOp(
 
 /// Performs const folding `calculate` with element-wise behavior on the two
 /// attributes in `operands` and returns the result if possible.
-template <class AttrElementT,
-          class ElementValueT = typename AttrElementT::ValueType,
-          class CalculationT =
-              std::function<ElementValueT(ElementValueT, ElementValueT)>>
-static Attribute constFoldBinaryOp(ArrayRef<Attribute> operands,
+template <class ArgAttrElementT, class RetAttrElementT = ArgAttrElementT,
+          class ArgElementValueT = typename ArgAttrElementT::ValueType,
+          class CalculationT = std::function<
+              ArgElementValueT(ArgElementValueT, ArgElementValueT)>>
+static Attribute constFoldBinaryOp(Operation *op, ArrayRef<Attribute> operands,
                                    const CalculationT &calculate) {
   assert(operands.size() == 2 && "binary op takes two operands");
-  if (auto lhs = operands[0].dyn_cast_or_null<AttrElementT>()) {
-    auto rhs = operands[1].dyn_cast_or_null<AttrElementT>();
+
+  assert(op->getNumResults() == 1 && "binary op needs one result");
+  auto returnType = op->getResult(0).getType();
+  // TODO(scotttodd): assert that RetAttrElementT matches returnType
+  //   or create an attribute using the general mlir::Type, without the extra
+  //   RetAttrElementT template type somehow
```

Comment:
I tried a few things here, but I couldn't find a way to either
A) get the required types at compile time from a single templated class or function argument
B) assert at runtime that the two sources of type data match

This is error-prone right now: you have to both add `<FloatAttr, IntegerAttr>` _and_ add a test exercising the code (with runtime crashes lacking clear error messages). I'd much prefer for the code to enforce those constraints directly.

Any suggestions? I suppose I could fork to a less general `constFoldBinaryComparisonOp` that just assumes `IntegerAttr` return, like the `constFoldCmpOp` (unary) function?

Reply:
returnType.isa<RetAttrElementT>() will tell you

Reply:
Tried that, `returnType` is a `mlir::Type` while `RetAttrElementT` is a `mlir::IntegerAttr` 

```
mlir/IR/StorageUniquerSupport.h(109): error C2338: casting from a non-convertible type
```

`IntegerAttribute` has a `getType()` function, but it isn't static (so need a concrete instance, not just the templated class name)

Reply:
I feel like there should be a simple way to do this and I'm holding attributes / types wrong, but maybe I could use the static `get()` and `verify()` methods, to exploit the class-specific code in `verify()`: https://github.com/llvm/llvm-project/blob/5492b71db3c4722afa39f7073996e9b5fb166d00/mlir/lib/IR/BuiltinAttributes.cpp#L357-L359